test/fake_http_server.cc

```diff
@@ -87,38 +98,73 @@ void FakeServer::TerminateAllStreams() {
 
 void FakeServer::Handler::operator()(Server::request const &request,
                                      Server::connection_ptr connection) {
-  auto stream_it = path_streams.find(request.destination);
-  if (stream_it != path_streams.end()) {
-    auto& stream = stream_it->second;
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-
-    // Create a queue for this watcher and add to the stream.
-    std::queue<std::string> my_queue;
-    stream.AddQueue(&my_queue);
-
-    // For every new string on my queue, send to the client.  The
-    // empty string indicates that we should terminate the stream.
-    while (true) {
-      std::string s = stream.GetNextResponse(&my_queue);
-      if (s.empty()) {
-        break;
+  if (request.method == "GET") {
+    auto stream_it = path_streams.find(request.destination);
+    if (stream_it != path_streams.end()) {
+      auto& stream = stream_it->second;
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+
+      // Create a queue for this watcher and add to the stream.
+      std::queue<std::string> my_queue;
+      stream.AddQueue(&my_queue);
+
+      // For every new string on my queue, send to the client.  The
+      // empty string indicates that we should terminate the stream.
+      while (true) {
+        std::string s = stream.GetNextResponse(&my_queue);
+        if (s.empty()) {
+          break;
+        }
+        connection->write(s);
       }
-      connection->write(s);
+      return;
+    }
+
+    auto it = get_handlers.find(request.destination);
+    if (it != get_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+      connection->write(handler(request.destination, headers));
+      return;
     }
-    return;
   }
 
-  auto it = path_responses.find(request.destination);
-  if (it != path_responses.end()) {
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-    connection->write(it->second);
-    return;
+  if (request.method == "POST") {
+    auto it = post_handlers.find(request.destination);
+    if (it != post_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      std::string body;
+      std::mutex mutex;
+      mutex.lock();
```

Comment:
Rather than locking the mutex explicitly, let's use RAII primitives (see, e.g., the [watch synchronization code](https://github.com/Stackdriver/metadata-agent/blob/master/src/kubernetes.cc#L817-L829) in `kubernetes.cc`).
Since move capture is not supported, you will have to capture by reference and explicitly `std::move` inside the lambda, e.g.:
```c++
      std::unique_lock<std::mutex> body_retrieved(body_mutex);
      connection->read([&](...) {
        std::unique_lock<std::mutex> done(std::move(body_retrieved));
        body.append(range.begin(), range.end());
      });
      std::lock_guard<std::mutex> await_body(body_mutex);
```

Reply:
Is it possible for the read callback to be called multiple times? If so, I'm not sure this suggestion would work as intended.

Reply:
This suggestion is equivalent to the explicit locking  code that it replaces. However, you have a good point — as the code was written originally (with the response in the callback), the synchronization code was consistent. But as written now, it does look like the callback was designed to be invoked more than once... We'd need to investigate how to ensure that there's only one invocation.

Reply:
The logic seems correct, but I don't like the raw unlock. Here are a few possible alternatives. My personal preference is for (1), but we can discuss.
1. Use a condition variable with a predicate (note that unlocking the mutex before notifying the condition variable is optional, so this could be made slightly simpler):
   ```c++
         std::mutex body_mutex;
         std::string body;
         std::condition_variable body_retrieved;
         std::unique_lock<std::mutex> body_lock(body_mutex);
         connection->read([&](Server::connection::input_range range,
                              boost::system::error_code error,
                              size_t size,
                              Server::connection_ptr conn) {
           {
             std::lock_guard<std::mutex> lock(body_mutex);
             body.append(range.begin(), range.end());
           }
           body_retrieved.notify_all();
         });
         body_retrieved.wait(body_lock, [&]() {
           return std::to_string(body.size()) == headers["Content-Length"];
         });
         connection->set_status(Server::connection::ok);
   ```
2. Use a condition variable (note that unlocking the mutex before notifying the condition variable is optional, so this could be made simpler):
   ```c++
         std::mutex body_mutex;
         std::string body;
         std::condition_variable body_retrieved;
         std::unique_lock<std::mutex> body_lock(body_mutex);
         connection->read([&](Server::connection::input_range range,
                              boost::system::error_code error,
                              size_t size,
                              Server::connection_ptr conn) {
           std::size_t body_size;
           {
             std::lock_guard<std::mutex> lock(body_mutex);
             body.append(range.begin(), range.end());
             body_size = body.size();
           }
           if (std::to_string(body_size) == headers["Content-Length"]) {
             body_retrieved.notify_all();
           }
         });
         body_retrieved.wait(body_lock);
         connection->set_status(Server::connection::ok);
   ```
3. Use a captured lock:
   ```c++
         std::mutex body_mutex;
         std::string body;
         std::unique_lock<std::mutex> body_retrieved(body_mutex);
         connection->read([&](Server::connection::input_range range,
                              boost::system::error_code error,
                              size_t size,
                              Server::connection_ptr conn) {
           body.append(range.begin(), range.end());
           if (std::to_string(body.size()) == headers["Content-Length"]) {
             body_retrieved.unlock();
           }
         });
         std::lock_guard<std::mutex> await_body(body_mutex);
         connection->set_status(Server::connection::ok);
   ```

Reply:
Thanks! Yeah I like (1) too.